kythe/go/serving/pipeline/nodes/nodes.go

```diff
@@ -0,0 +1,193 @@
+
+func init() {
+	beam.RegisterFunction(embedSourceKey)
+	beam.RegisterFunction(entryToNode)
+
+	beam.RegisterType(reflect.TypeOf((*combineNodes)(nil)).Elem())
+}
+
+// FromEntries transforms a PCollection of *ppb.Entry protos into *ppb.Nodes.
+func FromEntries(s beam.Scope, entries beam.PCollection) beam.PCollection {
+	return beam.ParDo(s, embedSourceKey,
+		beam.CombinePerKey(s, &combineNodes{},
+			beam.ParDo(s, entryToNode, entries)))
+}
+
+func entryToNode(e *spb.Entry, emit func(*spb.VName, *ppb.Node)) error {
+	if e.Source == nil {
+		return fmt.Errorf("invalid Entry: source is missing: %+v", e)
+	}
+
+	n := &ppb.Node{}
+	if e.EdgeKind == "" {
+		if e.FactName == "" || e.Target != nil {
+			return fmt.Errorf("invalid fact Entry: {%v}", e)
+		}
+
+		switch e.FactName {
+		case facts.NodeKind:
+			kind := string(e.FactValue)
+			if k := schema.NodeKind(kind); k != scpb.NodeKind_UNKNOWN_NODE_KIND {
+				n.Kind = &ppb.Node_KytheKind{k}
+			} else {
+				n.Kind = &ppb.Node_GenericKind{kind}
+			}
+		case facts.Subkind:
+			subkind := string(e.FactValue)
+			if k := schema.Subkind(subkind); k != scpb.Subkind_UNKNOWN_SUBKIND {
+				n.Subkind = &ppb.Node_KytheSubkind{k}
+			} else {
+				n.Subkind = &ppb.Node_GenericSubkind{subkind}
+			}
+		default:
+			n.Fact = append(n.Fact, entryToFact(e))
+		}
+	} else {
+		if (e.FactName != "/" && e.FactName != "") || len(e.FactValue) != 0 || e.Target == nil {
+			return fmt.Errorf("invalid edge Entry: {%v}", e)
+		}
+
+		n.Edge = append(n.Edge, entryToEdge(e))
+	}
+
+	emit(e.Source, n)
+	return nil
+}
+
+func entryToEdge(e *spb.Entry) *ppb.Edge {
+	kind, ord, _ := edges.ParseOrdinal(e.EdgeKind)
+	g := &ppb.Edge{Target: e.Target, Ordinal: int32(ord)}
+	edgeKind := schema.EdgeKind(kind)
+	if edgeKind == scpb.EdgeKind_UNKNOWN_EDGE_KIND {
+		g.Kind = &ppb.Edge_GenericKind{kind}
+	} else {
+		g.Kind = &ppb.Edge_KytheKind{edgeKind}
+	}
+	return g
+}
+
+func entryToFact(e *spb.Entry) *ppb.Fact {
+	f := &ppb.Fact{Value: e.FactValue}
+	name := schema.FactName(e.FactName)
+	if name == scpb.FactName_UNKNOWN_FACT_NAME {
+		f.Name = &ppb.Fact_GenericName{e.FactName}
+	} else {
+		f.Name = &ppb.Fact_KytheName{name}
+	}
+	return f
+}
+
+type combineNodes struct{}
+
+func (combineNodes) CreateAccumulator() *ppb.Node { return &ppb.Node{} }
+
+func (c *combineNodes) MergeAccumulators(accum, n *ppb.Node) *ppb.Node {
+	if n.Kind != nil {
+		accum.Kind = n.Kind
+	}
+	if n.Subkind != nil {
+		accum.Subkind = n.Subkind
+	}
+	for _, f := range n.Fact {
+		accum.Fact = append(accum.Fact, f)
+	}
+	for _, e := range n.Edge {
+		accum.Edge = append(accum.Edge, e)
+	}
+	return accum
+}
+
+func (c *combineNodes) AddInput(accum, n *ppb.Node) *ppb.Node { return c.MergeAccumulators(accum, n) }
+
+func (c *combineNodes) ExtractOutput(n *ppb.Node) *ppb.Node {
+	// TODO(schroederc): deduplicate earlier during combine
+	if len(n.Fact) > 1 {
+		sort.Slice(n.Fact, func(i, j int) bool { return compareFacts(n.Fact[i], n.Fact[j]) == compare.LT })
```

Comment:
nano-nit: I wouldn't use i and j both here and farther down in the method body (same for the Edge-related code below); using distinct variables would improve readability IMO.